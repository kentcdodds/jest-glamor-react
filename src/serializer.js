const {replaceSelectors} = require('./replace-selectors')
const {
  getNodes,
  markNodes,
  getSelectors,
  getStylesAndAllSelectors,
} = require('./utils')

function createSerializer(styleSheet, classNameReplacer) {
  function test(val) {
    return (
      val &&
      !val.serializedWithJestGlamorReact &&
      (val.$$typeof === Symbol.for('react.test.json') ||
        (val instanceof HTMLElement && !isBeingSerialized(val)))
    )
  }

  function print(val, printer) {
    const nodes = getNodes(val)
    markNodes(nodes)
    const selectors = getSelectors(nodes)
    const {styles, allSelectors} = getStylesAndAllSelectors(
      selectors,
      styleSheet,
    )
    const printedVal = printer(val)
    if (styles) {
      return replaceSelectors(
        allSelectors,
        styles,
        printedVal,
        classNameReplacer,
      )
    } else {
      return printedVal
    }
  }

  return {test, print}
}

function isBeingSerialized(node) {
  let currentNode = node

  while (currentNode) {
    if (currentNode.serializedWithJestGlamorReact) {
      return true
    }
    currentNode = currentNode.parentNode
  }
  return false
}

// doing this to make it easier for users to mock things
// like switching between development mode and whatnot.
const getGlamorStyleSheet = () => require('glamor').styleSheet
const glamorSerializer = createSerializer(getGlamorStyleSheet)
createSerializer.test = glamorSerializer.test
createSerializer.print = glamorSerializer.print

module.exports = createSerializer
