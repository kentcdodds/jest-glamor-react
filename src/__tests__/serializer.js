import React from 'react'
import renderer from 'react-test-renderer'
import * as glamor from 'glamor'
import * as enzyme from 'enzyme'
import toJson from 'enzyme-to-json'
import serializer from '../serializer'

expect.addSnapshotSerializer(serializer)

function Wrapper(props) {
  const className = glamor.css({
    padding: '4em',
    background: 'papayawhip',
  })
  return <section className={`${className}`} {...props} />
}

function Title(props) {
  const className = glamor.css({
    fontSize: '1.5em',
    textAlign: 'center',
    color: 'palevioletred',
  })
  return <h1 className={`${className}`} {...props} />
}

test('react-test-renderer', () => {
  const tree = renderer
    .create(
      <Wrapper>
        <Title>Hello World, this is my first glamor styled component!</Title>
      </Wrapper>,
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})

test('enzyme', () => {
  const ui = (
    <Wrapper>
      <Title>Hello World, this is my first glamor styled component!</Title>
    </Wrapper>
  )

  const enzymeMethods = ['shallow', 'mount', 'render']
  enzymeMethods.forEach(method => {
    const tree = enzyme[method](ui)
    expect(toJson(tree)).toMatchSnapshot(`enzyme.${method}`)
  })
})

test('works when the root element does not have styles', () => {
  const tree = renderer
    .create(
      <div>
        <Wrapper />
      </div>,
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})

test(`doesn't mess up stuff that does't have styles`, () => {
  const tree = renderer.create(<div />).toJSON()

  expect(tree).toMatchSnapshot()
})

test(`doesn't mess up stuff when styles have a child selector`, () => {
  const style = glamor.css({
    '> div': {
      display: 'inline-block',
    },
  })

  const tree = renderer
    .create(
      <div>
        <span {...style} />
      </div>,
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})

const generalTests = [
  {
    title: 'data attributes',
    styles: {
      backgroundColor: 'rebeccapurple',
      margin: 2,
    },
  },
  {
    title: 'media queries',
    styles: {
      fontSize: 24,
      margin: 12,
      '@media (max-width: 641px)': {
        fontSize: 20,
        margin: 10,
      },
    },
  },
  {
    title: 'pseudo elements',
    styles: {
      '& button': {
        color: 'green',
      },
    },
  },
  {
    title: 'pseudo states',
    styles: {
      ':hover': {
        backgroundColor: 'blue',
      },
    },
  },
  {
    title: 'Appended class',
    styles: {
      '&.button': {
        color: 'green',
      },
    },
  },
  {
    title: 'Child selector',
    styles: {
      '> div': {
        display: 'inline-block',
      },
    },
  },
]

generalTests.forEach(({title, styles}, index) => {
  test(title, () => {
    const tree = renderer.create(<div {...glamor.css(styles)} />)
    expect(tree).toMatchSnapshot(
      `${index + 1}. general tests: ${title} - ${JSON.stringify(styles)}`,
    )
  })
})
