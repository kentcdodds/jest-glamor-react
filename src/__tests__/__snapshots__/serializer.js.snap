// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Appended class: 5. general tests: Appended class - {"&.button":{"color":"green"}} 1`] = `
.css-0,
[data-css-1].button {
  color: green;
}

<div
  data-css-1=""
/>
`;

exports[`Child selector: 6. general tests: Child selector - {"> div":{"display":"inline-block"}} 1`] = `
.css-0> div,
[data-css-0]> div {
  display: inline-block;
}

<div
  data-css-0=""
/>
`;

exports[`data attributes: 1. general tests: data attributes - {"backgroundColor":"rebeccapurple","margin":2} 1`] = `
.css-0,
[data-css-0] {
  background-color: rebeccapurple;
  margin: 2px;
}

<div
  data-css-0=""
/>
`;

exports[`doesn't mess up stuff that does't have styles 1`] = `<div />`;

exports[`doesn't mess up stuff when styles have a child selector 1`] = `
.css-0> div,
[data-css-0]> div {
  display: inline-block;
}

<div>
  <span
    data-css-0=""
  />
</div>
`;

exports[`enzyme: enzyme.mount 1`] = `
.css-0,
[data-css-0] {
  padding: 4em;
  background: papayawhip;
}

.css-1,
[data-css-1] {
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
}

<Wrapper>
  <section
    className="css-0"
  >
    <Title>
      <h1
        className="css-1"
      >
        Hello World, this is my first glamor styled component!
      </h1>
    </Title>
  </section>
</Wrapper>
`;

exports[`enzyme: enzyme.render 1`] = `
.css-0,
[data-css-0] {
  padding: 4em;
  background: papayawhip;
}

.css-1,
[data-css-1] {
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
}

<section
  class="css-0"
>
  <h1
    class="css-1"
  >
    Hello World, this is my first glamor styled component!
  </h1>
</section>
`;

exports[`enzyme: enzyme.shallow 1`] = `
.css-0,
[data-css-0] {
  padding: 4em;
  background: papayawhip;
}

<section
  className="css-0"
>
  <Title>
    Hello World, this is my first glamor styled component!
  </Title>
</section>
`;

exports[`media queries: 2. general tests: media queries - {"fontSize":24,"margin":12,"@media (max-width: 641px)":{"fontSize":20,"margin":10}} 1`] = `
.css-0,
[data-css-0] {
  font-size: 24px;
  margin: 12px;
}

@media (max-width: 641px) {
  .css-0,
  [data-css-0] {
    font-size: 20px;
    margin: 10px;
  }
}

<div
  data-css-0=""
/>
`;

exports[`pseudo elements: 3. general tests: pseudo elements - {"& button":{"color":"green"}} 1`] = `
.css-0 button,
[data-css-0] button {
  color: green;
}

<div
  data-css-0=""
/>
`;

exports[`pseudo states: 4. general tests: pseudo states - {":hover":{"backgroundColor":"blue"}} 1`] = `
.css-0,
[data-css-1]:hover {
  background-color: blue;
}

<div
  data-css-1=""
/>
`;

exports[`react-test-renderer 1`] = `
.css-0,
[data-css-0] {
  padding: 4em;
  background: papayawhip;
}

.css-1,
[data-css-1] {
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
}

<section
  className="css-0"
>
  <h1
    className="css-1"
  >
    Hello World, this is my first glamor styled component!
  </h1>
</section>
`;

exports[`works when the root element does not have styles 1`] = `
.css-0,
[data-css-0] {
  padding: 4em;
  background: papayawhip;
}

<div>
  <section
    className="css-0"
  />
</div>
`;
